### PR TITLE
Battlegrounds maximum team size difference

### PR DESCRIPTION
**Pull request by BattleChickensRevenge**
Change maximum difference in team sizes from 1 to 0.

Where faction populations are not equal, there are from the larger faction always or almost always more players queuing, which means that the larger faction team is always or almost always as much larger than the small faction team as permitted by the maximum difference in team sizes.

A difference of only one player is enough for the larger team to always or almost always win.

On LH, Horde is the larger faction, on Darrow, Alliance, and it is seen that on LH, the Horde team is always or almost always one larger than the Alliance team, and vis-verse on Darrow.

This has a feedback effect, where the smaller faction always or almost always loses, and so its players stop queuing.

In short, even a difference of one means that BGs, by design, always or almost always lead one faction to win, consistently, on a per-server basis.

The existing server enforced team size difference of one was introduced presumably to ensure the teams are not so unbalanced that one team will inevitably win; but what has been done, with a limit of one, does not in fact achieve the intended purpose.  A maximum difference of zero will.

Please see issue 1071 for discussion.

https://github.com/LightsHope/server/issues/1071

### CODE REVIEW

**Comment by bigfgeg** [comment]
Shouldn't this be discussed with the project leads since this is clearly a custom balance change that is not addressing a bug of any sort?

**Comment by BattleChickensRevenge** [comment]
Utterly and absolutely, yes.

I have absolutely no idea how to go about this, though - how *do* you get the attention of the devs to discuss an issue?

Making a pull request served two purposes - it made the necessary change available, and brings the issue to the attention of the devs, so it can be considered.

One point though - I would be inclined to think of this as a bug, a *configuration* bug.

Functionality to limit team imbalance exists - there *is* a difference limiter in the code already - and it is configured to allow a difference of one.

It is reasonable to assume the purpose of the limiter is to prevent one team flattening the other - but in fact even a difference of one allows the larger team to always or almost always win, *and* where the more popular faction will generally be queuing more players, the *actual* outcome of current configuration is that one faction wins always or almost always, on a per-server basis.  This is obviously not what was intended.  It seems to be a configuration bug.

**Comment by brotalnia** [comment]
If it is a custom change that produces behavior differing from that of Blizzard's servers, then it must be a config option and it must be at blizzlike value by default.

**Comment by BattleChickensRevenge** [comment]
This pull request will not make the change a config option.

Making it a config option can be done.  I'll need to figure out how config options are provied.

An interesting questions here is what the Blizz-like value is, and if it changed over time : if Bliz recognized the problem and adjusted the value, it would be hard to know, because the consequences of the change are impossible or nearly impossible to find in the historical record.

**Comment by knasen88** [comment]
It was made with some leniency I presume so the BGqueue would pop sooner rather than later. This would make people being slow to join game push the time until the BG is full by alot, because it needs to wait on every queuing player to accept or drop queue before it can pop for the guy next in queue. You want both teams full before the match starts, joining late is not fun. If every player queuing would join with LP autojoin at last second, you would probably have upwards of 10 minutes until the BG is full.

And if you find your team constantly underpopulated, it's probably cause you're playing alliance non peak hours, most BG's i join are full within a minute or two unless i'm playing at really weird hours.

**Comment by knasen88** [comment]
They are issued serially, if you ever play an AV "prem" where you que together only about 15-20 gets the pop first, then as more people join more people get the que. If they weren't what do you think would happen to people in a premade accepting ques late cos worldpvp/whatver? get shuffled into another game because some pug insta-accepted and stole his spot?

Do not mess with the current because this is literally a non issue, non 60 BG's will forever be a joke even if not balanced teams

**Comment by BattleChickensRevenge** [comment]
> If they
> weren't what do you think would happen to people in a premade accepting ques late cos 
> worldpvp/whatver? get shuffled into another game because some pug insta-accepted 
> and stole his spot?

I think I understand.

The problem here is actually the *design* of BG invitations - you queue, and then you are invited, and you can accept or decline the invitiation.

Imagine we have a new, empty BG.  We have say five Alliance players queued (premade makes no difference, I think - we'll see) and say seven Horde.

Now, if queuing mean you would accept automatically, we'd have no problem; we'd just teleport the five Alliance player and five of the Horde players into the BG.  If new people queue, then we accept new people such that the team numbers in game remain balanced (so if say one player quits and it's now 4-5, then we'd accept a single Alliance player next).

Easy.

But - ah - that's *not* how it works.

What happens is that players queue and then they are offered an invitation, which they can accept or decline.  If they ignore it, it declines after a little while - about 30 seconds.

So now we have a problem if we want to keep teams balanced - imagine the teams are balanced, say 5-5.  We then see we have an Alliance player queued and a Horde player queued.  Great!  we issue invites to both.

Oh.  What happens if one of them now declines his bloody invite?  what happens is we quite likely have accepted the other player into the game and so now the teams are unbalanced - and if we've in fact said that our maximum team difference is *zero*, well, now we're completely stuck because the BG invitation design as just can't work with a difference of zero - how can we issued invites which *might be declined* and keep teams *always* balanced?  that's actually imossible.

The problem with having a difference of one - which is needed by the current design - is that is *also* means one facton always or almost always wins on a per-server basis.  Unintended consequences being as ever much more profound than the *intended* consequence.

(So far, out of eight BG games, the teams were full - and so equal - in one, and Alliance wone.  Horde won the other seven games.)

> Do not mess with the current because this is literally a non issue, non 60 BG's will
> forever be a joke even if not balanced teams

Why?

I've been playing them for over a year.  They matter to me and I'd like them to work properly.

**Comment by BattleChickensRevenge** [comment]
So, given the current BG design, a permitted team size difference of one is mandatory.

However, the unintended consequence of this necessity is that the more populous faction always or almost always has a larger team - the permitted maximum difference is in fact the *usual* difference.

There is I think a way we can improve matters.

When teams are equal, the server only issues an invite to players if I can issue the same number of invites to both factions.

Right now, if the teams are equal, say 5-5, and a Horde player queues, he will get an invite, because the server is fine with 5-6 (difference of one).

What I propose is that when teams are equal, 5-5, invites will only be issued if the same number of invites can be issued to both factions.  Typically then this means both an Alliance player and a Horde player have queued.  We will then normally go 5-5 to 6-6, 6-6 to 7-7, etc.  What happens *now* is we go to 5-6, then 6-7, then 7-8, etc.

It may be then that one of these players declines and the teams are unbalanced, but then the server will only issue an invite to the weaker team faction, so the server is here doing the right thing and trying to get the teams balanced again.
